edb/server/protocol/auth_ext/smtp.py

```diff
@@ -0,0 +1,152 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2023-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from typing import *
+
+import asyncio
+import email
+import os
+import pickle
+
+import aiosmtplib
+
+from edb.common import retryloop
+from edb.ir import statypes
+
+from . import util
+
+
+_semaphore: asyncio.BoundedSemaphore | None = None
+
+
+async def send_email(
+    db: Any,
+    message: Union[
+        email.message.EmailMessage,
+        email.message.Message,
+        str,
+        bytes,
+    ],
+    sender: Optional[str] = None,
+    recipients: Optional[Union[str, Sequence[str]]] = None,
+    test_mode: bool = False,
+) -> None:
+    global _semaphore
+    if _semaphore is None:
+        _semaphore = asyncio.BoundedSemaphore(
+            int(os.environ.get("EDGEDB_SERVER_AUTH_SMTP_CONCURRENCY", 5))
+        )
+
+    host = util.maybe_get_config(
+        db.db_config,  # type: ignore
+        "ext::auth::SMTPConfig::host",
+    ) or "localhost"
+    port = util.maybe_get_config(
+        db.db_config,  # type: ignore
+        "ext::auth::SMTPConfig::port",
+        expected_type=int,
+    )
+    username = util.maybe_get_config(
+        db.db_config,  # type: ignore
+        "ext::auth::SMTPConfig::username",
+    )
+    password = util.maybe_get_config(
+        db.db_config,  # type: ignore
+        "ext::auth::SMTPConfig::password",
+    )
+    timeout_per_attempt = util.get_config(
+        db.db_config,  # type: ignore
+        "ext::auth::SMTPConfig::timeout_per_attempt",
+        expected_type=statypes.Duration,
+    )
+    req_timeout = timeout_per_attempt.to_microseconds() / 1_000_000.0
```

Comment:
`req_timeout` is measured in millionths of a microsecond?

Reply:
Yes I believe so: https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-INTERVAL-INPUT

Reply:
Ah, should use `*` - good catch!